tests/phpunit/tests/base-sender.php

```diff
@@ -3,74 +3,67 @@
 
-		if ( is_string( $sender ) ) {
-			$sender_instance = new WPNotify_BaseSender( $sender );
-		} else if ( is_array( $sender ) ) {
-			$sender_instance = new WPNotify_BaseSender( $sender['name'], $sender['image'] );
-		}
-
-		$sender_encoded = json_encode( $sender_instance );
+		$sender_encoded = json_encode( $sender );
 
 		$this->assertEquals( $expected_json, $sender_encoded );
 	}
 
 	/**
-	 * @param string|array $sender
-	 * @param string       $json
+	 * @param array  $sender_params
+	 * @param string $json
 	 *
 	 * @dataProvider data_provider_senders
 	 */
-	public function test_it_can_be_instantiated_from_json( $sender, $json ) {
+	public function test_it_can_be_instantiated_from_json( $sender_params, $json ) {
 
 		$testee = WPNotify_BaseSender::json_unserialize( $json );
 
-		$sender_name     = '';
-		$sender_instance = null;
-
-		if ( is_string( $sender ) ) {
-			$sender_name     = $sender;
-			$sender_instance = new WPNotify_BaseSender( $sender );
-		} else if ( is_array( $sender ) ) {
-			$sender_name     = $sender['name'];
-			$sender_instance = new WPNotify_BaseSender( $sender['name'], $sender['image'] );
-		}
+		$sender_reflection = new ReflectionClass( 'WPNotify_BaseSender' );
+		/** @var WPNotify_BaseSender $sender */
+		$sender = $sender_reflection->newInstanceArgs( array_values( (array) $sender_params ) );
 
 		$this->assertInstanceOf( 'WPNotify_BaseSender', $testee );
-		$this->assertEquals( $sender_name, $testee->get_name() );
+		$this->assertEquals( $sender->get_name(), $testee->get_name() );
+
+		$this->assertEquals( $sender->get_image(), $testee->get_image() );
 
 		/**
 		 * If an Image has been sent
 		 */
 		if ( $testee->get_image() ) {
-			$this->assertInstanceOf( 'WPNotify_BaseImage', $testee->get_image() );
-			$this->assertEquals( $sender_instance->get_image()->get_source(), $testee->get_image()->get_source() );
-			$this->assertEquals( $sender_instance->get_image()->get_alt(), $testee->get_image()->get_alt() );
+			$this->assertInstanceOf( get_class( $sender->get_image() ), $testee->get_image() );
+			$this->assertEquals( $sender->get_image()->get_source(), $testee->get_image()->get_source() );
+			$this->assertEquals( $sender->get_image()->get_alt(), $testee->get_image()->get_alt() );
 		}
 	}
 
 	public function data_provider_senders() {
+
 		return array(
 
-			array(
-				'Name 1',
-				'{"name":"Name 1","image":null}',
+			'sender without image' => array(
+				array(
+					'Name 1'
+				),
+				'{"name":"Name 1"}',
 			),
 
-			array(
+			'sender with image' => array(
 				array(
-					'name'  => 'Name 2',
-					'image' => new WPNotify_BaseImage( 'img-source', 'img-alt' ),
+					'name'               => 'Name 2',
+					'WPNotify_BaseImage' => new WPNotify_BaseImage( 'img-source', 'img-alt' ),
 				),
-				'{"name":"Name 2","image":{"source":"img-source","alt":"img-alt"}}',
+				'{"name":"Name 2","WPNotify_BaseImage":{"source":"img-source","alt":"img-alt"}}',
```

Comment:
The Json serialization was meant to be used to communicate between the back-end and front-end parts of the feature.

I am not sure what the front-end part would do with a PHP class name. Wouldn't this be bothersome to have to change the data field names on the front-end if we ever want to add a new PHP class for images?

Reply:
This is already the way `WPNotify_BaseNotifications` are instantiated, it's better to keep our code consistent.